### ISSUE TITLE
Fix S3267 FP: `ref struct` types cannot leave the stack

### ISSUE THREAD

**Issue by fiotti** [SonarSource/sonar-dotnet#8430]
### Description

Some suggestions to simplify the code by rewriting it using LINQ cannot be applied when `ref struct` types are involved.

### Repro steps

```csharp
class SonarTest
{
    void Visit(ReadOnlySpan<char> x)
    {
        // ...
    }

    void Test(IEnumerable<ReadOnlyMemory<char>> tokens)
    {
        // S3267: loop should be simplified by calling Select(token => token.Span))
        foreach (ReadOnlyMemory<char> token in tokens)
        {
            ReadOnlySpan<char> chars = token.Span;
            Visit(chars);
        }
    }
}
```

Unfortunately, `ReadOnlySpan<char>` is a `ref struct` type and it can't leave the stack, so it can't be returned from the selector passed to `Select()`.

```csharp
// CS0306: The type 'ReadOnlySpan<char>' may not be used as a type argument
foreach (ReadOnlySpan<char> chars in tokens.Select(t => t.Span))
{
    Visit(chars);
}
```

### Expected behavior

I didn't expect any diagnostics to be emitted.

### Actual behavior

False positive.

### Known workarounds

```csharp
#pragma warning disable S3267
```

### Related information

* SonarLint for Visual Studio 2022, version 7.4.0.80741
* Visual Studio Professional 2022 (64 bit), version 17.7.6
* dotnet 7.0.403
* MSBuild version 17.7.2+d6990bcfa for .NET Framework, version 17.7.2.37605
* Windows 11 Pro 22H2 22621.2715

**Comment by gregory-paidis-sonarsource** [comment]
Hey there.
I can confirm that this is an FP.
I added a [reproducer](https://github.com/SonarSource/sonar-dotnet/pull/8455) to track this.
Thanks for raising this very interesting FP!

**Comment by abj-proton** [comment]
Encountered this too. Any plans to fix it?

**Comment by martin-strecker-sonarsource** [comment]
Another report from @Corniel was filed here:
https://community.sonarsource.com/t/137678

**Comment by mary-georgiou-sonarsource** [comment]
Fixed in sonar-dotnet 10.9
